resolwe/storage/connectors/transfer.py

```diff
@@ -0,0 +1,245 @@
+                ):
+                    self.__buffer_modify.wait(CircularBuffer.DEFAULT_READ_WAIT_TIMEOUT)
+                bytes_to_read = min(self.__bytes_available(), size)
+                slice_start = self.__tail
+                slice_end = slice_start + bytes_to_read
+                if slice_end <= self.__bs:
+                    ret[ret_pos : ret_pos + bytes_to_read] = self.__buffer[
+                        slice_start:slice_end
+                    ]
+                else:  # rotation
+                    slice_end -= self.__bs
+                    sep = self.__bs - slice_start
+                    ret[ret_pos : ret_pos + sep] = self.__buffer[
+                        slice_start : self.__bs
+                    ]
+                    ret[ret_pos + sep : ret_pos + sep + slice_end] = self.__buffer[
+                        0:slice_end
+                    ]
+                self.__tail = slice_end
+                ret_pos += bytes_to_read
+                self._bytes_read += bytes_to_read
+                self.__buffer_modify.notify()  # Notify the writing thread
+                if self.closed:
+                    break
+
+            self.__reading_bytes = 0
+            self.__buffer_modify.notify()  # Notify the writing thread
+            # print("reading: OUT!!!!!!!!!!!!!!!")
+            return ret[0:ret_pos].tobytes()
+
+    def write(self, data: bytes) -> int:
+        """Write data into the stream and return the number of bytes actually
+        written.
+        """
+        datamv = memoryview(data)
+        data_pos = 0
+        with self.__buffer_modify:
+            # Wait for enough bytes to become available
+            while data_pos < len(datamv):
+                while not self.__can_write:
+                    self.__buffer_modify.wait()
+                bytes_to_write = min(self.__bytes_free(), len(datamv) - data_pos)
+                slice_start = self.__head
+                slice_end = slice_start + bytes_to_write
+                if slice_end <= self.__bs:
+                    self.__buffer[slice_start : slice_start + bytes_to_write] = datamv[
+                        data_pos : data_pos + bytes_to_write
+                    ]
+                else:  # rotation
+                    slice_end %= self.__bs
+                    sep = self.__bs - slice_start
+                    self.__buffer[slice_start : self.__bs] = datamv[
+                        data_pos : data_pos + sep
+                    ]
+                    remaining_bytes = bytes_to_write - sep
+                    self.__buffer[0:remaining_bytes] = datamv[
+                        data_pos + sep : data_pos + bytes_to_write
+                    ]
+                self.__head += bytes_to_write
+                self.__head %= self.__bs
+                data_pos += bytes_to_write
+                if self.__reading_bytes > 0:
+                    if self.__bytes_available() >= self.__reading_bytes:
```

Comment:
I would prefer to notify the other thread anyway, and it can decide whether to proceed or to wait. This way, you better encapsulate the reading logic - it is all in a single function, and `__reading_bytes` variable doesn't need to be defined on the object level.

Reply:
That was the first implementation. But since google was reading in 100M chunks and S3 writing in 64k chunks, I has around 1500 notification per chunk. And that was way slower that this implementation.

Reply:
Ok, let's keep it this way then.